### DISCUSSION TITLE
Subtleness & details of index algorithm setup on data

### DISCUSSION THREAD

**Discussion by molokanov50** [milvus-io/milvus#5528] [Q&A and General discussion] [answered]
Hi team,
The subject of my interest is whether I can setup Milvus' index algorithms on my data, apart from simply specifying user parameters in my `index_param` block (such as `nlist` for IVF indexes). Is there any means to perform user-defined learning on the data? 
E.g., let's take some of Faiss' IVF algorithms. Their coarse quantizer is somehow trained on the collection. Is there an opportunity to influence this training by specifying additional parameters or by performing supervised learning? Is there an opportunity to select any part of my collection, or this quantizer can be trained only on the whole collection? Is data sampling implemented somehow, and what does data sampling depend on?

**Comment by yhmo** [comment, marked as answer]
Hi @molokanov50 
The milvus client api create_index() has a parameter named "params", it is a json format parameter. So, the client can send user defined parameter for index. But to parse user defined parameters and let them take effect, you need to modify the source code of search engine.

Milvus splits data into segments, you can refer this page: https://milvus.io/docs/v1.1.0/storage_concept.md. Milvus build index for each segment, not for whole collection. Currently each segment size has limitation(up to 4GB), to build index on whole collection, you need to modify source code to remove the limitation.

To create data sampling, you need to add more source code in search engine, there are two ways to output sampling data:
1. Print sampling data into log file as text format(need to write some source code).
2. Milvus has an internal command between client and server, named "cmd()", this api can send a sting to server, ask a string return. You can use this api to return sampling data(also need to write some source code).

**Reply by molokanov50** [reply]
Hi @yhmo 
And does Milvus have an opportunity to index only one partition of a collection, not the whole collection?

**Reply by yhmo** [reply]
No, currently no api to do this.

**Reply by molokanov50** [reply]
@yhmo 
Thanks for your clarifications. And you said that Milvus builds its index for each segment. Does this mean that the mentioned coarse quantizer (which is essentially k-means for IVFs) is trained for each segment separately and independently, so that each segment receives its own `nlist` clusters (with their own centroids not coinciding with the centroids of another segment's clusters), and the actual number of clusters in my collection is `nlist * number_of_segments`? Or still on the contrary, `nlist` is the total number of clusters in the whole collection?

**Reply by yhmo** [reply]
In milvus, index is trained for each segment separately and independently.

Assume there is a collection contains 3 segments. You create an index IVFs, nlist=1000, each segment has an index file, there will be 3 index files generated. Each index file has 1000 clusters, trained by its segment data.
Then receive a search request, nq=1, nprobe=5, topk=10. Milvus search the index files one by one, with same parameter(nprobe=5, topk=10). For each index, return a result contains nq * topk pairs(ids to distances). So there are 3 results output from search engine. Then milvus reduce the 3 results into one final result, the final result also contains nq * topk pairs. The client receive the final result.

For your questions: 
Each segment generate an index contains nlist clusters, so, for the whole collection, there are nlist * number_of_segments clusters.
